src/main/java/com/google/cloud/tools/appengine/v2/internal/process/CliProcessManager.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.appengine.v2.internal.process;
+
+import com.google.cloud.tools.appengine.v2.OutputHandler;
+import com.google.cloud.tools.appengine.v2.internal.process.io.StringResultConverter;
+import com.google.common.util.concurrent.ListenableFutureTask;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.util.Scanner;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * The cli process mechanism takes a process and starts 3 threads on an executor to manage it
+ * 1. The processMain thread waits on the process to end and stores the exit code
+ * 2. The StdErr thread processes the err stream and writes output to the outputListener
+ * 3. The StdOut thread processes the out stream and stores output to return as result
+ * The future implementation considers the process main (1) thread to be the primary thread
+ * to watch. It forwards all future interface calls to the future returned when starting that
+ * thread (1).
+ *
+ * @param <T> the process return type
+ */
+public class CliProcessManager<T> implements Future<T> {
+
+  private final ExecutorService executor;
+  private final OutputHandler outputHandler;
+  private final StringResultConverter<T> stringResultConverter;
+  private final Process process;
+  private ListenableFutureTask<CliProcessResult<T>> processMain;
+  private ListenableFutureTask<String> processStdOut;
+  private Runnable processStdErr;
+
+  private CliProcessManager(Process process, OutputHandler outputHandler,
+      StringResultConverter<T> stringResultConverter) {
+
+    this.process = process;
+    this.executor = MoreExecutors.listeningDecorator(MoreExecutors.getExitingExecutorService(
+        (ThreadPoolExecutor) Executors.newFixedThreadPool(3), 2, TimeUnit.SECONDS));
+    this.outputHandler = outputHandler;
+    this.stringResultConverter = stringResultConverter;
+  }
+
+  // Main entry point, adds three managing threads to the executor
+  private Future<T> manage() {
+
+    processStdOut = ListenableFutureTask.create(new Callable<String>() {
```

Comment:
@joao Random question: I noticed that the IJ servers framework was strongly
geared towards managing executing of Application Servers via start and stop
shell scripts. For IntelliJ at least, do you think it could simplify things
just to use the common library to help produce the shell command rather
than to execute the development server itself?



On Wed, Aug 24, 2016 at 2:30 PM, João Martins <notifications@github.com>
wrote:

> In src/main/java/com/google/cloud/tools/appengine/v2/internal/process/
> CliProcessManager.java
> <https://github.com/GoogleCloudPlatform/appengine-plugins-core/pull/182#discussion_r76112602>
> :
>
> > +  private Runnable processStdErr;
> > +
> > +  private CliProcessManager(Process process, OutputHandler outputHandler,
> > +      StringResultConverter<T> stringResultConverter) {
> > +
> > +    this.process = process;
> > +    this.executor = MoreExecutors.listeningDecorator(MoreExecutors.getExitingExecutorService(
> > +        (ThreadPoolExecutor) Executors.newFixedThreadPool(3), 2, TimeUnit.SECONDS));
> > +    this.outputHandler = outputHandler;
> > +    this.stringResultConverter = stringResultConverter;
> > +  }
> > +
> > +  // Main entry point, adds three managing threads to the executor
> > +  private Future<T> manage() {
> > +
> > +    processStdOut = ListenableFutureTask.create(new Callable<String>() {
>
> We might want to not specify any listeners. That is the case in IntelliJ,
> where we want the process to have its default stdout/stderr streams so that
> the framework can listen there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-plugins-core/pull/182/files/3f589b068bc9778596fd487d60abd976feaf8ce5#r76112602>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HbKrhQR_74nxhlZKSFsQIdOlUC2_ks5qjI2wgaJpZM4JYOIg>
> .
>